modules/videoio/src/cap.cpp

```diff
@@ -371,45 +371,56 @@ CV_IMPL CvCapture * cvCreateFileCapture (const char * filename)
 CV_IMPL CvVideoWriter* cvCreateVideoWriter( const char* filename, int fourcc,
                                             double fps, CvSize frameSize, int is_color )
 {
-    // If none of the writers is used
-    // these statements suppress 'unused parameter' warnings.
+    return cvCreateVideoWriterWithPreference(CV_CAP_ANY, filename, fourcc, fps, frameSize, is_color);
+}
+
+CvVideoWriter* cvCreateVideoWriterWithPreference(int apiPreference, const char* filename, int fourcc,
+                                            double fps, CvSize frameSize, int is_color )
+{
     CV_UNUSED(frameSize);
     CV_UNUSED(is_color);
 
-    //CV_FUNCNAME( "cvCreateVideoWriter" );
-
     CvVideoWriter *result = 0;
 
     if(!fourcc || !fps)
         TRY_OPEN(result, cvCreateVideoWriter_Images(filename))
 
-#ifdef HAVE_FFMPEG
-    TRY_OPEN(result, cvCreateVideoWriter_FFMPEG_proxy (filename, fourcc, fps, frameSize, is_color))
-#endif
-
-#ifdef HAVE_VFW
-    TRY_OPEN(result, cvCreateVideoWriter_VFW(filename, fourcc, fps, frameSize, is_color))
-#endif
-
-#ifdef HAVE_MSMF
-    TRY_OPEN(result, cvCreateVideoWriter_MSMF(filename, fourcc, fps, frameSize, is_color))
-#endif
-
-/*  #ifdef HAVE_XINE
-    TRY_OPEN(result, cvCreateVideoWriter_XINE(filename, fourcc, fps, frameSize, is_color))
-    #endif
-*/
-#ifdef HAVE_AVFOUNDATION
-    TRY_OPEN(result, cvCreateVideoWriter_AVFoundation(filename, fourcc, fps, frameSize, is_color))
-#endif
-
-#if defined(HAVE_QUICKTIME) || defined(HAVE_QTKIT)
-    TRY_OPEN(result, cvCreateVideoWriter_QT(filename, fourcc, fps, frameSize, is_color))
-#endif
-
-#ifdef HAVE_GSTREAMER
-    TRY_OPEN(result, cvCreateVideoWriter_GStreamer(filename, fourcc, fps, frameSize, is_color))
-#endif
+    switch(apiPreference)
+    {
+        default:
+            //exit if the specified API is unavaliable
+            if (apiPreference != CV_CAP_ANY) break;
+        case(CV_CAP_FFMPEG):
```

Comment:
In my opinion wrapping whole _case_ blocks into _#ifdef..#endif_ will look more clean:
```.cpp
#ifdef HAVE_FFMPEG
        case CV_CAP_FFMPEG:
            TRY_OPEN(result, cvCreateVideoWriter_FFMPEG_proxy (filename, fourcc, fps, frameSize, is_color))
            if (apiPreference != CV_CAP_ANY) break;
#endif

#ifdef HAVE_VFW
        case CV_CAP_VFW:
            TRY_OPEN(result, cvCreateVideoWriter_VFW(filename, fourcc, fps, frameSize, is_color))
            if (apiPreference != CV_CAP_ANY) break;
#endif
```